include/LightGBM/config.h

```diff
@@ -954,6 +954,12 @@ struct Config {
   // desc = set this to ``true`` to use double precision math on GPU (by default single precision is used)
   bool gpu_use_dp = false;
 
+#ifdef USE_CUDA
+  // desc = number of gpus (CUDA implementation only) LGBM_CUDA
+  // desc = default value is 1
+  int num_gpu = 1;
+#endif
```

Comment:
this also breaks the auto-generated config. 
maybe you can remove `#ifdef USE_CUDA`  block, and check num_gpu when enable CUDA.

Reply:
Please explain your comment(s) about auto-generated config or codes.  What does this mean?

Removing the #ifdef USE_CUDA is not a problem in the config.cpp and .h files.

Reply:
@ChipKerchner

> Please explain your comment(s) about auto-generated config or codes. What does this mean?

Please refer to the note at the beginning of `config-auto.cpp` file:
https://github.com/microsoft/LightGBM/blob/1c21d12eb392d589ee6c255e286c2437725f1361/src/io/config_auto.cpp#L5-L6

We generate that file automatically from `config.h`. That is why `config-auto.cpp` should be never modified manually and `config.h` should contain only very basic parameter definitions with comments. All manual checks should be done inside `config.cpp` file.

Reply:
I'm seeing checks all over the place in config_auto.cpp (outside the changes we have done)

Reply:
Do you mean those `CHECK_*`s?
https://github.com/microsoft/LightGBM/blob/61b3c30842810205bc95736e6e66baa790b098cd/src/io/config_auto.cpp#L438-L439
Those ones are generated automatically.

Reply:
Yes.  How are those checks "generated automatically"?

Reply:
From comments in `config.h` file. For example,
https://github.com/microsoft/LightGBM/blob/61b3c30842810205bc95736e6e66baa790b098cd/include/LightGBM/config.h#L308-L309

And here is the corresponding code:
https://github.com/microsoft/LightGBM/blob/61b3c30842810205bc95736e6e66baa790b098cd/helpers/parameter_generator.py#L175-L179

Reply:
How does the new "config_auto.cpp" get generated?

Reply:
Simply by running `python helpers/parameter_generator.py` after modifying `config.h`.